### PR TITLE
Open Enum Transform for AWS

### PR DESCRIPTION
**Pull request by lewisjkl**
Adds a transform which makes all AWS enums open as we've seen that they frequently need to be treated as such. Going to add a scripted test before moving this PR out of draft status.

### CODE REVIEW

File: modules/codegen-plugin/src/sbt-test/codegen-plugin/aws-open-enums/src/main/scala/Main.scala [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+import com.amazonaws.dynamodb.TestEnum
+
+object Main extends App {
+
+  if (TestEnum.hints.get[alloy.OpenEnum].isEmpty)
+    sys.error("Expected OpenEnum trait on aws enum.")
```
Comment (kubukoz) [inline modules/codegen-plugin/src/sbt-test/codegen-plugin/aws-open-enums/src/main/scala/Main.scala]: **nitpick:** could this be a `require(TestEnum.hints.has(alloy.OpenEnum), "Expected ...")`?

File: modules/codegen/src/smithy4s/codegen/transformers/OpenEnumTransformer.scala [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+@annotation.nowarn("msg=class EnumTrait in package traits is deprecated")
+private[codegen] final class OpenEnumTransformer extends ProjectionTransformer {
+  override def getName: String = OpenEnumTransformer.name
+
+  private val awsNamespacePrefix = "com.amazonaws"
+
+  def transform(ctx: TransformContext): Model = {
+    val shapeMapper: Function[Shape, Shape] = { (shp: Shape) =>
+      if (shp.toShapeId.getNamespace.startsWith(awsNamespacePrefix)) {
```
Comment (kubukoz) [inline modules/codegen/src/smithy4s/codegen/transformers/OpenEnumTransformer.scala]: **nitpick:** you can replace the `if` and pattern match with a pattern match on the top level of the lambda

```scala
= {
  case shp if !shp.getId.getNamespace.startsWith(...) => shp
  case e: EnumShape => ...
  case e: IntEnumShape => ...
  ...
}
```

File: modules/codegen/src/smithy4s/codegen/transformers/OpenEnumTransformer.scala [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        shp match {
+          case e: EnumShape =>
+            val _ = e.hasTrait(classOf[EnumTrait])
```
Comment (kubukoz) [inline modules/codegen/src/smithy4s/codegen/transformers/OpenEnumTransformer.scala]: **question:** what does this do?

File: modules/codegen/src/smithy4s/codegen/transformers/OpenEnumTransformer.scala [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+          case e: IntEnumShape =>
+            e.toBuilder.addTrait(new OpenEnumTrait()).build()
+          case t: Shape with ToSmithyBuilder[_]
+              if t.hasTrait(classOf[EnumTrait]) =>
+            t.toBuilder() match {
+              case b: AbstractShapeBuilder[_, _] =>
+                b.addTrait(new OpenEnumTrait())
+                b.build().asInstanceOf[Shape]
+              case _ => t
+            }
```
Comment (kubukoz) [inline modules/codegen/src/smithy4s/codegen/transformers/OpenEnumTransformer.scala]: **suggestion:** how do you feel about this?

```scala
case t: Shape if t.hasTrait(classOf[EnumTrait]) =>
  (Shape
    .shapeToBuilder(t): AbstractShapeBuilder[_, _])
    .addTrait(new OpenEnumTrait())
    .build()
```

File: modules/codegen/test/src/smithy4s/codegen/transformers/OpenEnumTransformerSpec.scala [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        .transform(
+          TransformContext
+            .builder()
+            .model(loadModel(input))
+            .build()
+        )
+
+    val containsOpenEnum = transformedModel
+      .getShape(ShapeId.fromParts("test", "Test"))
+      .get
```
Comment (kubukoz) [inline modules/codegen/test/src/smithy4s/codegen/transformers/OpenEnumTransformerSpec.scala]: **nitpick:** `getShape + get` -> `expectShape`

File: modules/codegen/test/src/smithy4s/codegen/transformers/OpenEnumTransformerSpec.scala [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+    val containsOpenEnum = transformedModel
+      .getShape(ShapeId.fromParts("com.amazonaws.kinesis", "Test"))
+      .get
+      .hasTrait(classOf[OpenEnumTrait])
+
+    assert(containsOpenEnum, "Expected openEnum trait to be present")
+  }
+
+  private def loadModel(namespaces: String*): Model = {
```
Comment (kubukoz) [inline modules/codegen/test/src/smithy4s/codegen/transformers/OpenEnumTransformerSpec.scala]: **nitpick:** `namespaces` isn't super accurate, maybe `modelFiles`?

File: modules/codegen/src/smithy4s/codegen/transformers/OpenEnumTransformer.scala (hunk 3 above)
Comment (lewisjkl) [reply modules/codegen/src/smithy4s/codegen/transformers/OpenEnumTransformer.scala]: Accidentally left that there, thanks for spotting

**Review by Baccata — APPROVED** [review]

**Pull request merged**